src/dns.rs

```diff
@@ -1,58 +1,85 @@
 mod local;
+mod default;
+mod cache;
 
 use crate::types::Result;
 use std::net::IpAddr;
-use crate::dns::local::LocalOverrideTable;
-use crate::types;
+
+/// A single IP entry in the DNS cache
+struct IpEntry {
+    /// Actual IP address (either ipv4 or ipv6)
+    ip: IpAddr,
+    /// Epoch when ip address expires
+    expired: u64,
+    /// Epoch when this entry was last used
+    last_used: u64,
+}
 
 /// A DNS entry is a simple domain name to IP address mapping
+#[derive(Default)]
 struct DNSEntry {
+    // domain name
     domain_name: String,
-    ip: IpAddr,
+    // List of IPv4 addresses for this domain name
+    ipv4: Vec<IpEntry>,
+    // List of IPv6 addresses for this domain name
+    ipv6: Vec<IpEntry>,
+    // Round robin pointer in case we need to round robin
+    rr_ipv4_ptr: usize,
+    // Round robin pointer in case we need to round robin
+    rr_ipv6_ptr: usize,
 }
 
-/// The main DNS resolver struct
-struct DnsResolver {
-    /// Local override table enabled
-    local_override_enabled: bool,
-    /// Local override table
-    local_override: LocalOverrideTable,
-    /// DNS over HTTPS enabled
-    doh_enabled: bool,
-    /// DNS over TLS enabled
-    dot_enabled: bool,
-    /// TTL override enabled
-    ttl_override_enabled: bool,
-    /// TTL override in seconds
-    ttl_override: u32,
-    /// List of DNS resolvers that are used for resolving domains
-    resolvers: Vec<IpAddr>,
+impl DNSEntry {
+    pub(crate) fn new(domain: &str) -> DNSEntry {
+        DNSEntry {
+            domain_name: domain.to_string(),
+            ..Default::default()
+        }
+    }
 }
 
-impl DnsResolver {
-    pub fn new() -> Self {
-        DnsResolver {
-            local_override_enabled: false,
-            local_override: LocalOverrideTable::default(),
-            doh_enabled: false,
-            dot_enabled: false,
-            ttl_override_enabled: false,
-            ttl_override: 0,
-            resolvers: Vec::new(),
-        }
+impl DNSEntry {
+    /// Returns true when this entry has a ipv4 address
+    fn has_ipv4(&self) -> bool {
+        !self.ipv4.is_empty()
     }
 
-    pub fn resolve(&mut self, domain: &str) -> Result<IpAddr> {
-        if self.local_override_enabled {
-            if let Some(ip) = self.local_override.resolve(domain) {
-                return Ok(ip);
-            }
-        }
+    /// Returns true when this entry has a ipv6 address
+    fn has_ipv6(&self) -> bool {
+        !self.ipv6.is_empty()
+    }
+
+    /// Retrieves the next ipv4 address in the round robin list
+    fn next_ipv4(&mut self) -> IpAddr {
+        let entry = self.ipv4.get(self.rr_ipv4_ptr).expect("Invalid round robin pointer");
```

Comment:
I think, the expect shouldn't be here. This error is recoverable (just ignore the ip ptr), so maybe just use a Result or an Option / "silently" recover the error may be better.